File: Infinite Scroll (ObjC)/InfiniteScrollSample/ArticleViewController.m

#import "ArticleViewController.h"

@import NativoSDK;

@interface ArticleViewController () <WKNavigationDelegate, NtvSectionDelegate>
@property (nonatomic) UIActivityIndicatorView *loadingView;
@property (weak, nonatomic) IBOutlet UIView *nativoAdView;
@property (weak, nonatomic) IBOutlet WKWebView *webView;
@property (weak, nonatomic) IBOutlet UIScrollView *scrollView;

@end

NSString * const SECTION_URL = @"nativo.net/bottomofarticle";


@implementation ArticleViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    if (self.articleURL) {
        self.webView.navigationDelegate = self;
        [self.webView loadRequest:[NSURLRequest requestWithURL:self.articleURL]];
        self.loadingView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
        self.loadingView.center = self.webView.center;
        [self.view addSubview:self.loadingView];
        [self.loadingView startAnimating];
    }
}

- (void)dealloc {
    [self.webView stopLoading];
}

#pragma mark - WKWebViewNavigation

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    [self.loadingView stopAnimating];
    [self.webView expandWebViewToFitContents];
    
    // Place Nativo ad in view when webview has finished loading
    if (self.nativoAdView.subviews.count == 0) {
        [NativoSDK setSectionDelegate:self forSection:SECTION_URL];
        [NativoSDK placeAdInView:self.nativoAdView atLocationIdentifier:self.articleURL inContainer:self.scrollView forSection:SECTION_URL options:nil];
    }
}

#pragma mark - NtvSectionDelegate Methods

- (NSString *)section:(NSString *)sectionUrl registerNibNameForAdTemplateType:(NtvAdTemplateType)templateType atLocationIdentifier:(id)locationIdentifier {
    switch (templateType) {
        case NtvAdTemplateTypeNative:
            return @"ArticleNativeAdView";
            
        case NtvAdTemplateTypeVideo:
            return @"ArticleVideoAdView";
            
        default:
            return nil;
    }
}

- (void)section:(NSString *)sectionUrl needsReloadDatasourceAtLocationIdentifier:(id)identifier forReason:(NSString *)reason {
    
    if ([reason isEqualToString:NtvSectionReloadReasonRemoveView]) {
        
        // Set height to 0 instead of remove (If removed from parent view you will need to reapply autolayout constraints)
        [self.nativoAdView.heightAnchor constraintEqualToConstant:0].active = YES;
        NSLog(@"Removing Nativo ad view");
    }
}

- (void)section:(NSString *)sectionUrl needsDisplayLandingPage:(UIViewController<NtvLandingPageInterface> *)sponsoredLandingPageViewController {
    [self.navigationController pushViewController:sponsoredLandingPageViewController animated:YES];
}

- (void)section:(NSString *)sectionUrl needsDisplayClickoutURL:(NSURL *)url {
    // Load click-out url in WKWebView
    UIViewController *clickoutAdVC = [[UIViewController alloc] init];
    WKWebView *webView = [[WKWebView alloc] initWithFrame:clickoutAdVC.view.frame];
    [clickoutAdVC.view addSubview:webView];
    NSURLRequest *clickoutReq = [NSURLRequest requestWithURL:url];
    [webView loadRequest: clickoutReq];
    [self.navigationController pushViewController:clickoutAdVC animated:YES];
}


@end

File: Table View (Swift)/NativoSwiftTableViewSample/ArticleViewController.swift
//
//  ArticleViewController.swift
//  NativoSwiftTableViewSample
//
//  Copyright © 2019 Nativo. All rights reserved.
//

import UIKit
import NativoSDK

class ArticleViewController: UIViewController {
    
    @IBOutlet weak var webView: WKWebView!
    @IBOutlet weak var nativoAdView: UIView!
    @IBOutlet weak var scrollView: UIScrollView!
    
    var articleURL : URL?
    let spinner = UIActivityIndicatorView(style: .gray)
    let ArticleNativoSectionUrl = "www.publisher/bottom-of-article"

    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.addSubview(spinner)
        self.spinner.center = self.view.center
        self.spinner.startAnimating()
        
        NativoSDK.setSectionDelegate(self, forSection: ArticleNativoSectionUrl)
        
        if let url = self.articleURL {
            self.webView.navigationDelegate = self;
            self.webView.load(URLRequest(url: url))
        }
    }
}

extension ArticleViewController: WKNavigationDelegate {
    
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        self.spinner.stopAnimating()
        self.webView.expandToFitContents()
                
        // Place bottom-of-article Nativo ad in view when webview has finished loading
        if self.nativoAdView.subviews.count == 0 {
            NativoSDK.placeAd(in: self.nativoAdView, atLocationIdentifier: self.articleURL as Any, inContainer: self.scrollView, forSection: ArticleNativoSectionUrl, options: nil)
        }
    }
}

extension ArticleViewController: NtvSectionDelegate {
    
    func section(_ sectionUrl: String, needsPlaceAdInViewAtLocation identifier: Any) {
        
    }
    
    func section(_ sectionUrl: String, needsRemoveAdViewAtLocation identifier: Any) {
        // Set height to 0 instead of remove (If removed from parent view you will need to reapply autolayout constraints)
        self.nativoAdView.widthAnchor.constraint(equalToConstant: 0).isActive = true
        print("Removed Nativo ad view")
    }
    
    func section(_ sectionUrl: String, registerNibNameFor templateType: NtvAdTemplateType, atLocationIdentifier locationIdentifier: Any) -> String? {
        switch templateType {
            
        case .native:
            return "ArticleCell"
        
        case .video:
            return "ArticleVideoAdCell"
            
        default:
            return nil
        }
    }

    func section(_ sectionUrl: String, needsDisplayLandingPage sponsoredLandingPageViewController: (UIViewController & NtvLandingPageInterface)?) {
        if let landingPage = sponsoredLandingPageViewController {
            self.navigationController?.pushViewController(landingPage, animated: true)
        }
    }
    
    func section(_ sectionUrl: String, needsDisplayClickoutURL url: URL) {
        let clickoutAdVC = UIViewController()
        let webView = UIWebView(frame: clickoutAdVC.view.frame)
        clickoutAdVC.view.addSubview(webView)
        let clickoutReq = URLRequest(url: url)
        webView.loadRequest(clickoutReq)
        self.navigationController?.pushViewController(clickoutAdVC, animated: true)
    }
}
